crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,777 @@
+        }
+    }
+
+    // none of the order's tokens fulfilled any of the prefered denominations
+    // so just pick the first one if there was any success at all
+    if !apy_denominations.is_empty() {
+        order_apy.apy = Some(apy_denominations[0].clone());
+    }
+
+    Ok(order_apy)
+}
+
+/// Calculates each token vault apy at the given timeframe
+pub fn get_token_vaults_apy(
+    trades: &[Trade],
+    vols: &[VaultVolume],
+    start_timestamp: Option<u64>,
+    end_timestamp: Option<u64>,
+) -> Result<Vec<TokenVaultAPY>, OrderbookSubgraphClientError> {
+    let mut token_vaults_apy: Vec<TokenVaultAPY> = vec![];
+    for vol in vols {
+        // this token vault trades in desc order by timestamp
+        let vault_trades = trades
+            .iter()
+            .filter(|v| {
+                (v.input_vault_balance_change.vault.vault_id.0 == vol.id
+                    && v.input_vault_balance_change.vault.token == vol.token)
+                    || (v.output_vault_balance_change.vault.vault_id.0 == vol.id
+                        && v.output_vault_balance_change.vault.token == vol.token)
+            })
+            .collect::<Vec<&Trade>>();
+
+        // this token vault first trade, indictaes the start time
+        // to find the end of the first day to find the starting capital
+        let first_trade = vault_trades[vault_trades.len() - 1];
+        let first_day_last_trade = vault_trades
+            .iter()
+            .filter(|v| {
+                u64::from_str(&v.timestamp.0).unwrap()
+                    <= u64::from_str(&first_trade.timestamp.0).unwrap() + DAY
+            })
+            .collect::<Vec<&&Trade>>()[0];
+
+        // vaults starting capital at end of first day of its first ever trade
+        let starting_capital = if first_day_last_trade
+            .input_vault_balance_change
+            .vault
+            .vault_id
+            .0
+            == vol.id
+            && first_day_last_trade.input_vault_balance_change.vault.token == vol.token
+        {
+            U256::from_str(
+                &first_day_last_trade
+                    .input_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        } else {
+            U256::from_str(
+                &first_day_last_trade
+                    .output_vault_balance_change
+                    .new_vault_balance
+                    .0,
+            )?
+        };
+
+        // the time range for this token vault
+        let mut start = u64::from_str(&first_trade.timestamp.0)?;
+        start_timestamp.inspect(|t| {
+            if start > *t {
+                start = *t;
+            }
+        });
+        let end = end_timestamp.unwrap_or(chrono::Utc::now().timestamp() as u64);
+
+        // this token vault apy
+        let apy = if starting_capital.is_zero() {
+            0_f64
+        } else {
+            // by 4 point decimals
```

Comment:
why not arbitrary precision? or at least 18 decimals which everything else is normalized to?

Reply:
well `f64` cant cover full 18 decimals point, but can go with a `int` such as `I256` with 18 point decimals as the value, is that what you mean?